# SLR - StochasticGD (FAKE DATA)

In [ ]:
import numpy as np

def simple_linear_regression_sgd(X, y, lr=0.01, epochs=1000):
    m = 0
    b = 0
    for _ in range(epochs):
        for i in range(len(X)):
            y_pred = m * X[i] + b
            error = y_pred - y[i]
            m -= lr * error * X[i]
            b -= lr * error
    return m, b


np.random.seed(42)
X_simple = 2 * np.random.rand(100, 1)
y_simple = 4 + 3 * X_simple + np.random.randn(100, 1)
m_simple, b_simple = simple_linear_regression_sgd(X_simple.flatten(), y_simple.flatten())
print(f"Simple Linear Regression: m = {m_simple}, b = {b_simple}")

Simple Linear Regression: m = 2.817038877249641, b = 4.240691077501467


# MLR - StochasticGD (FAKE DATA)

In [ ]:
import numpy as np

def multiple_linear_regression_sgd(X, y, lr=0.01, epochs=1000):
    m = np.zeros(X.shape[1])
    b = 0
    for _ in range(epochs):
        for i in range(len(X)):
            y_pred = np.dot(X[i], m) + b
            error = y_pred - y[i]
            m -= lr * error * X[i]
            b -= lr * error
    return m, b

np.random.seed(42)
X_multiple = 2 * np.random.rand(100, 2)
y_multiple = 4 + 3 * X_multiple[:, 0] + 5 * X_multiple[:, 1] + np.random.randn(100)
m_multiple, b_multiple = multiple_linear_regression_sgd(X_multiple, y_multiple)
print(f"Multiple Linear Regression: m = {m_multiple}, b = {b_multiple}")

Multiple Linear Regression: m = [3.15953153 5.16015879], b = 3.7453025538856846


# SLR - StochasticGD (KAGGLE SAMPLE DATASET)

In [ ]:
import numpy as np
import pandas as pd

df = pd.read_csv('/content/sample_data/Salary_dataset.csv')
def simple_linear_regression_sgd(X, y, lr=0.01, epochs=1000):
    m = 0
    b = 0
    for _ in range(epochs):
        for i in range(len(X)):
            y_pred = m * X[i] + b
            error = y_pred - y[i]
            m -= lr * error * X[i]
            b -= lr * error
    return m, b

X_simple = df['YearsExperience'].values
y_simple = df['Salary'].values
m_simple, b_simple = simple_linear_regression_sgd(X_simple.flatten(), y_simple.flatten())
print(f"Simple Linear Regression: m = {m_simple}, b = {b_simple}")

Simple Linear Regression: m = 8894.994273837554, b = 27230.69426420553



## **Comparison: SGD Linear Regression vs Elastic Net (SGD)**

| Method                                       | Slope (m) | Intercept (b) |
| -------------------------------------------- | --------- | ------------- |
| **Simple Linear Regression (SGD)**           | 8894.99   | 27230.69      |
| **Elastic Net Regularized Regression (SGD)** | 8772.52   | 28446.03      |

###**Observation**

* The **slope decreased** slightly after applying Elastic Net regularization (from **8894 → 8772**), meaning the model becomes **less sensitive to the feature**.
* The **intercept increased** (from **27230 → 28446**), adjusting the line upward to balance the reduced slope.
* This indicates that **Elastic Net penalty pulled the coefficient towards zero**, achieving a controlled shrinkage effect.

###**Conclusion**

* **SGD without regularization** purely fits data and may risk overfitting.
* **Elastic Net SGD** introduces a balance of **L1 (sparsity)** and **L2 (stability)**, resulting in **more controlled and generalized coefficients**, ideal when expecting noisy or correlated data.



# MLR - StochasticGD (KAGGLE SAMPLE DATASET)

In [ ]:
import numpy as np
import pandas as pd

def multiple_linear_regression_sgd(X, y, lr=0.0000001, epochs=1000):
    m = np.zeros(X.shape[1])
    b = 0
    for _ in range(epochs):
        for i in range(len(X)):
            y_pred = np.dot(X[i], m) + b
            error = y_pred - y[i]
            m -= lr * error * X[i]
            b -= lr * error
    return m, b

df = pd.read_csv('/content/sample_data/Student_Performance.csv')

X_multiple = df[['Hours Studied', 'Previous Scores', 'Sleep Hours', 'Sample Question Papers Practiced']].values
y_multiple = df['Performance Index'].values
m_multiple, b_multiple = multiple_linear_regression_sgd(X_multiple, y_multiple)
print(f"Multiple Linear Regression: m = {m_multiple}, b = {b_multiple}")

Multiple Linear Regression: m = [ 2.18897517  0.82009164 -1.45585384 -0.26887853], b = -0.9208991215793743



## **Comparison: Multiple Linear Regression (SGD) vs Manual Elastic Net (SGD)**

| Method                               | θ₀ (Intercept) | θ₁ (Hours Studied) | θ₂ (Previous Scores) | θ₃ (Sleep Hours) | θ₄ (Sample Papers) |
| ------------------------------------ | -------------- | ------------------ | -------------------- | ---------------- | ------------------ |
| **Multiple Linear Regression (SGD)** | -0.92          | 2.1889             | 0.8201               | -1.4559          | -0.2689            |
| **Manual Elastic Net (SGD, α=0.1)**  | -33.22         | 2.8053             | 1.0252               | 0.4423           | 0.1858             |

###**Observation**

* **Intercept:** Elastic Net moved the intercept strongly negative (-33.22), showing the regularization adjusting the baseline.
* **Slopes:**

  * Elastic Net **increased positive contributions** (θ₁, θ₂)
  * Elastic Net **reversed negative contributions** (θ₃, θ₄) compared to standard SGD.
* Regularization **shrinks or adjusts weights** to reduce overfitting and balance contributions of features.

###**Conclusion**

* **SGD Linear Regression:** Purely fits the dataset; may reflect original correlations, including negative coefficients.
* **Elastic Net (Manual SGD):** Introduces **L1 and L2 penalties**, adjusting coefficients toward more generalized values; some weak or negative feature effects are reduced or reversed.
* Elastic Net is more suitable when expecting **correlated features or noisy data**.
